Add Boss Room Architecture (#879)
Unity-Technologies/com.unity.multiplayer.docs: docs/learn/bossroom/architecture.md

See the [Multiplayer over the internet](getting-started-boss-room.md) section of the Boss Room README for more information about using the two network transport mechanisms.

Boss Room uses the Unity Transport Package (UTP). Boss Room's assigns its instance of Unity Transport to the `transport` field of the `NetworkManager`.

The Unity Transport Package is a network transport layer with network simulation tools that help spot networking issues early during development. The UTP protocol initializes using direct IP connections but switches to Unity Relay at runtime if it detects that the client (or host) is using the Lobby service.

Unity Transport supports Unity Relay (provided by Unity Gaming Services). See the documentation on [Unity Transport Package](../../../transport/about.md) and [Unity Relay](https://docs-multiplayer.unity3d.com/docs/relay/relay) for more information.

## Connection flow state machine

> The `ConnectionManager`, a simple state machine, owns Boss Room’s network connection flow. It receives inputs from NGO (or the user) and handles the inputs according to its current state. Each state inherits from the C`onnectionState` abstract class. If you add a new transport, you must extend the `StartingHostState` and `ClientConnectingState` states. Both of these classes assume you're using the UTP transport.

Reviewer: ```suggestion
The `ConnectionManager`, a simple state machine, owns Boss Room’s network connection flow. It receives inputs from NGO (or the user) and handles the inputs according to its current state. Each state inherits from the `ConnectionState` abstract class. If you add a new transport, you must extend the `StartingHostState` and `ClientConnectingState` states. Both of these classes assume you're using the UTP transport.
```

Author: Nice catch 😄

* `ClientCharacter` is primarily a host for the `ClientActionPlayer` class and has the client RPCs for the character gameplay logic.

### Game config setup

We defined Boss Room’s game configuration using ScriptableObjects.

The [`GameDataSource.cs`](https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/blob/main/Assets/Scripts/Gameplay/GameplayObjects/RuntimeDataContainers/GameDataSource.cs) singleton class stores all actions and character classes in the game.

[`CharacterClass`](https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/blob/main/Assets/Scripts/Gameplay/Configuration/CharacterClass.cs) is the data representation of a Character and has elements such as starting stats and a list of Actions the character can perform. It covers both player characters and non-player characters (NPCs) alike.

The [`Action`](https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/blob/main/Assets/Scripts/Gameplay/Configuration/Action.cs) subclasses are data-driven and represent discrete verbs, such as swinging a weapon or reviving a player.

### Action System

:::note

> We built Boss Room's action system specifically for Boss Room's educational purpose. You'll need to implement a custom action system to allow for better game design emergence from your game designers.

Reviewer: ```suggestion
We built Boss Room's action system specifically for Boss Room's educational purpose. You'll need to implement a more user friendly custom action system to allow for better game design emergence from your game designers.
```

It’s also possible to play an anticipatory animation on the client requesting an ability. For instance, you might want to play a small jump animation when the character receives movement input but hasn’t yet moved (because it hasn’t synchronized the data from the server).

[`Server`](https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/blob/main) and [`Client`](https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/blob/main) `ActionPlayers` are companion classes to actions Boss Room uses to play out the actions on both the client and server.

#### Movement action flow

The following list describes the movement flow of a player character.

1. The client uses a mouse to click on the target destination.
2. The client sends an RPC (containing the target destination) to the server.
3. Anticipatory animation plays immediately on the client.
4. There’s network latency before the server receives the RPC.
5. The server receives the RPC (containing the target destination).
6. The server performs pathfinding calculations.

> 7. The server completes the pathfinding, and the server representation of the entity starts updating its `NetworkVariables` at the same cadence as `FixedUpdate`.

Reviewer: ```suggestion
7. The server completes the pathfinding, and the server representation of the entity starts updating its `NetworkTransform` at the same cadence as `FixedUpdate`.
```

To solve this challenge, we explored different client-server code separation approaches. We eventually decided to revert our initial client/server/shared assemblies to a more classic domain-driven assembly architecture while keeping more complex classes separated by client/server.

Our initial thought was that separating assemblies by client and server would allow for easier porting to a dedicated game server (DGS) model afterward; we’d only need to strip a single assembly to ensure code only runs when necessary.

However, this approach had two primary issues:

* It introduced callback hell, which made code (that should be trivial) too complex. You can look at our different action implementations in Boss Room 1.3.1 to see this.
* Many components could be single simple classes instead of three-class horrors.

After investigation, we determined that client/server separation was unnecessary for the following reasons:

* There’s no need for [asmdef (assembly definition)](https://docs.unity3d.com/Manual/cus-asmdef.html) stripping because you can ifdef out single classes instead.

> * It’s not completely necessary to ifdef classes because it’s only compile-time insurance that certain parts of client-side code never run.

Reviewer: ```suggestion
* It’s not completely necessary to ifdef classes because it’s only compile-time insurance that certain parts of client-side code never run. You can still disable the component on Awake at runtime if it's not mean to run on the server or client.
```

After those experiments, we established new rules for the team:

* Use domain-based assemblies
* Use single classes for small components (for example, the Boss Room door with a simple on/off state).
* If a class never grows too big, use a single NetworkBehaviour (because it’s easy to maintain).
* Use client and server classes (with each pointing to the other) for client/server separation.
* Place client/server pairs in the same assembly.
* If you start the game as a client, the server components disable themselves, leaving you with only client components executing. Ensure you don’t destroy the server components. NGO still requires the server component for network message sending.
* The clients have a `m_Server`, and servers have a `m_Client` property.
* The `Server` class owns server-driven `NetworkVariables`. Similarly, the `Client` class owns owner-driven `NetworkVariables`. This ownership separation helps make larger classes more readable and maintainable.

> * Use partial classes when separating ownership isn’t possible.

Reviewer: ```suggestion
* Use partial classes when separating by context isn’t possible.
```

### Publisher-Subscriber Messaging

We implemented a dependency injection-friendly Publisher-Subscriber pattern that allows Boss Room to send and receive strongly-typed messages in a loosely-coupled manner, where communicating systems only know about the `IPublisher`/`ISubscriber` of a given message type.

Because publishers and subscribers are classes, we can have more interesting behavior for message transfer, such as buffered and networked messaging. See the NetworkedMessageChannel section.

These mechanisms allow Boss Room to avoid circular references and limit the dependency surface between assemblies. Cross-communicating systems rely on common messages but don't necessarily need to know about each other. Because messages don’t need to know about each other, we can more easily separate them into smaller assemblies.

These mechanisms allow for strong separation of concerns and coupling reduction using PubSub and dependency injection (DI). We use DI to pass the handles to the `IPublisher` or `ISubscriber` of any given event type. As a result, message publishers and consumers are unaware of each other.

> `MessageChannel` classes implement the `IPublisher` and `ISubscriber` interfaces and provide the actual messaging logic.
> 
> #### `NetworkedMessageChannel`

Reviewer: ```suggestion
`MessageChannel` classes implement the `IPublisher` and `ISubscriber` interfaces and provide the actual messaging logic.

:::note
Boss Room could have used a 3rd party library for messaging (like MessagePipe). Since we needed custom networked channels and our use cases were quite simple, we decided to go with our own light implementation. If we had had a full game with more gameplay, we might have switched to a library that had more features.
:::

#### `NetworkedMessageChannel`
```
